llama_index/indices/postprocessor/relevancy/base.py

```diff
@@ -0,0 +1,214 @@
+                "join_on": "...",
+            },
+        ]
+    },
+}
+"""
+
+
+class BaseRelevancyPostprocessor(BaseNodePostprocessor):
+    """
+    Base class for relevancy post-processors.
+
+    It is meant to encompass the stages of:
+    - Splitting / Chunking
+    - Ranking Function / Relevancy Scoring
+    - Filtering / Marginalization
+    - Node Synthesis: Aggregation, Mapping
+    """
+
+    filter_operator_shuffle: OperatorMode
+    scorer_operator_shuffle: OperatorMode = filter_operator_shuffle
+
+    _initial_nodes: Optional[List[NodeWithScores]] = None
+    _nodes: Optional[List[NodeWithScores]] = None
+    _subnodes: Optional[List[List[NodeWithScores]]] = None
+
+    filters = List[BaseFilter]
+    scorer: BaseRelevancyScorer  # default is to use old score...?
+
+    def __init__(self, config):
+        self.config = config
+
+    def validate(self):
+        if self.splitter is None:
+            assert self.scorer.agg_mode == AggMode.INTER_NODE
+            for filter in self.filters:
+                assert filter.agg_mode == AggMode.INTER_NODE
+            for synthesizer in self.node_synthesizers:
+                assert synthesizer.op_mode == MapMode.NODE
+        else:
+            assert self.scorer.agg_mode != AggMode.INTER_NODE
+            for filter in self.filters:
+                assert filter.agg_mode != AggMode.INTER_NODE
+            for synthesizer in self.node_synthesizers:
+                assert synthesizer.op_mode in [
+                    MapMode.SUB_NODE,
+                    MapMode.NODE,
+                    AggMode.INTRA_NODE,
+                ]
+
+    def postprocess_nodes(
+        self,
+        nodes: List[NodeWithScore],
+        query_bundle: Optional[QueryBundle] = None,
+    ) -> List[NodeWithScore]:
+        self._initial_nodes = nodes
+
+        if self.scorer.agg_mode == AggMode.INTER_NODE:
+            assert self.splitter is None
+            self._nodes = nodes
+
+            # Relevant scores are inter-node
+            self._nodes = self.relevancy_scorer(self._nodes)
+
+            # Filter is inter-node
+            assert self.filter.agg_mode == AggMode.INTER_NODE
+            self._nodes = self.filter.filter(self._nodes)
+
+            # Synthesizer can only be a map from each node to a new node
+            assert self.synthesizer.op_mode == MapMode.NODE
+            self._nodes = self.synthesizer.synthesize(self._nodes)
+        else:
+            assert self.splitter is not None
+            self._subnodes = self.splitter.split(nodes)
+
+            # Score each subnode
+            if self.scorer.agg_mode == AggMode.INTRA_NODE:
+                relevancy_scored = []
+                for nodes in self._subnodes:
+                    relevancy_scored.append(self.scorer.score(nodes))
+                self._subnodes = relevancy_scored
+            elif self.scorer.agg_mode == AggMode.CROSS_NODE:
+                self._subnodes = self.scorer.score_nested(self._subnodes)
```

Comment:
maybe rename these methods `score_cross_node`, `filter_cross_node` to be more consistent.